File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/ReverseKGroup.java
package com.only4.algorithm.version4j.leetcode;

import com.only4.algorithm.version4j.extra.ListNode;

public class ReverseKGroup {
    public ListNode reverseKGroup(ListNode head, int k) {
        if (head == null || k == 1) return head;

        ListNode dummy = new ListNode(0);
        dummy.next = head;

        int totalLength = getLength(head);
        int groupCount = totalLength / k;

        ListNode prevGroupTail = dummy;
        ListNode currentGroupHead = head;

        // 进行groupCount次反转
        for (int i = 0; i < groupCount; i++) {
            ListNode nextGroupHead = getKthNode(currentGroupHead, k).next;

            // 反转当前组，返回反转后的新头节点
            ListNode reversedGroupHead = reverseKNodes(currentGroupHead, k);

            // 连接反转后的组
            prevGroupTail.next = reversedGroupHead;
            currentGroupHead.next = nextGroupHead;

            // 更新指针，准备处理下一组
            prevGroupTail = currentGroupHead;
            currentGroupHead = nextGroupHead;
        }

        return dummy.next;
    }

    private int getLength(ListNode head) {
        int length = 0;
        while (head != null) {
            head = head.next;
            length++;
        }
        return length;
    }

    private ListNode getKthNode(ListNode head, int k) {
        ListNode current = head;
        for (int i = 1; i < k && current != null; i++) {
            current = current.next;
        }
        return current;
    }

    private ListNode reverseKNodes(ListNode head, int k) {
        ListNode prev = null;
        ListNode current = head;

        for (int i = 0; i < k && current != null; i++) {
            ListNode next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }

        return prev; // 返回反转后的头节点
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/reverseKGroup.kt
package com.only4.algorithm.version4k.leetcode

import com.only4.algorithm.version4k.extra.ListNode

fun reverseKGroup(head: ListNode?, k: Int): ListNode? {
    if (head == null || k == 1) return head

    fun getLength(): Int {
        var length = 0
        var current = head
        while (current != null) {
            current = current.next
            length++
        }
        return length
    }

    fun getKthNode(head: ListNode, k: Int): ListNode {
        var current = head
        repeat(k - 1) {
            current = current.next!!
        }
        return current
    }

    fun reverseKNodes(head: ListNode, k: Int): ListNode {
        var prev: ListNode? = null
        var current: ListNode? = head

        repeat(k) {
            val next = current!!.next
            current.next = prev
            prev = current
            current = next
        }

        return prev!! // 返回反转后的头节点
    }

    val dummy = ListNode(0)
    dummy.next = head

    val totalLength = getLength()
    val groupCount = totalLength / k

    var prevGroupTail = dummy
    var currentGroupHead = head

    // 进行groupCount次反转
    repeat(groupCount) {
        val nextGroupHead = getKthNode(currentGroupHead!!, k).next

        val reversedGroupHead = reverseKNodes(currentGroupHead, k)

        // 连接反转后的组
        prevGroupTail.next = reversedGroupHead
        currentGroupHead.next = nextGroupHead

        // 更新指针，准备处理下一组
        prevGroupTail = currentGroupHead
        currentGroupHead = nextGroupHead
    }

    return dummy.next
}
